### PR TITLE
Fix/powi overflow and tighter integers

### PR DESCRIPTION
**Pull request by arthurvinci**
This PR solves overflow errors in the `powi` function. The following code would overflow at every line involving  `powi`:
```Rust
    #[test]
    fn test_powi_max_decimal() {
        let _max_bug = Decimal::MAX.powi(1);
        let _max_sqrt = Decimal::MAX.sqrt().unwrap();
        let _max_cbrt = Decimal::MAX.cbrt();
        let _max_dec_2 = _max_sqrt.powi(2);
        let _max_dec_3 = _max_cbrt.powi(3);
    }
    ```
    
    The PR also changes integer used in root function of `Decimal` and `PreciseDecimal` to tighter integers (I384 and I768 as suggested by @r001)

### CODE REVIEW

File: radix-engine-interface/src/math/decimal.rs [hunk 1]
```diff
@@ -177,24 +177,30 @@ impl Decimal {
...
-            return Decimal(
-                &base * Decimal(&base * &base / &one).powi(div(sub(exp, 1), 2)).0 / &one,
-            );
+            let dec_256: I256 = I256::try_from(&base_384 * &base_384 / &one_384).expect("Overflow");
+            return Decimal(dec_256).powi(div(exp, 2));
+        }
+        else {
+            let sub_dec_256: I256 = I256::try_from(&base_384 * &base_384 / &one_384).expect("Overflow");
+            let sub_dec = Decimal(sub_dec_256);
+            return Decimal(self.0) * sub_dec.powi(div(sub(exp, 1), 2));
```
Comment (r001) [inline radix-engine-interface/src/math/decimal.rs:203]: Please change `Decimal(self.0)` to `*self`.

File: radix-engine-interface/src/math/decimal.rs [hunk 2]
```diff
@@ -177,24 +177,30 @@ impl Decimal {
...
         if exp < 0 {
-            return Decimal(&one * &one / base).powi(mul(exp, -1));
+            let dec_256: I256 = I256::try_from(&one_384 * &one_384 / &base_384).expect("Overflow");
+            return Decimal(dec_256).powi(mul(exp, -1));
         }
         if exp == 0 {
             return Self::ONE;
         }
+        if exp == 1 {
+            return Decimal(self.0);
```
Comment (r001) [inline radix-engine-interface/src/math/decimal.rs:194]: Please change `Decimal(self.0)` to `*self`! Thanks

**Review by r001 — CHANGES_REQUESTED** [review]

**Comment by r001** [comment]
Another great PR to make Scrypto even safer. 🎉 Thanks @arthurvinci! 
Highly appreciate your work!

**Review by r001 — APPROVED** [review]

**Review by r001 — CHANGES_REQUESTED** [review]
Please rebase this on `betanet-v1` branch instead of develop.
Thank you, otherwise its all approved.

**Comment by arthurvinci** [comment]
@r001 Done. I am not sure it is doing what you expected though.

**Comment by r001** [comment]
Not really. I think what you should do is just clone the `betanet-v1` and copy the files you changed on top of it. 
If you can force `-f` push here, that would be great, if not, then create a new PR.

**Comment by r001** [comment]
@arthurvinci this is not as simple unfortunately as to just set the branch to `betanet-v1`. Your original branch was probably `develop`, which currently differ from `betanet-v1` as we are adding different changes to both. So what I suggest you to do is : 

1. Please pull `betanet-v1`.
2. Issue your changes made to files, rename branch to `fix/powi_overflow_and_tighter_integers`.
3. Force push to `arthurvinci:fix/powi_overflow_and_tighter_integers`.
4. If 3. does not work, then create a new PR, I will accept it. 
5. Done merged.

**Comment by arthurvinci** [comment]
@r001 This is what I did in #693 maybe I should have used a better name though.

**Comment by arthurvinci** [comment]
#693 solves the issues and makes required changes

**Pull request closed without merging**